pkg/querier/tripperware/shard_by_query.go

```diff
@@ -0,0 +1,81 @@
+package tripperware
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/thanos-io/thanos/pkg/querysharding"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+
+	cquerysharding "github.com/cortexproject/cortex/pkg/querysharding"
+)
+
+func ShardByMiddleware(logger log.Logger, limits Limits, merger Merger, numShards int) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQueryByMiddleware{
+			next:      next,
+			limits:    limits,
+			merger:    merger,
+			logger:    logger,
+			numShards: numShards,
+		}
+	})
+}
+
+type shardQueryByMiddleware struct {
+	next          Handler
+	limits        Limits
+	logger        log.Logger
+	merger        Merger
+	queryAnalyzer *querysharding.QueryAnalyzer
+	numShards     int
+}
+
+func (s shardQueryByMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	analysis, err := s.queryAnalyzer.Analyze(r.GetQuery())
+	if err != nil {
+		return nil, err
+	}
+
+	if !analysis.IsShardable() {
+		return s.next.Do(ctx, r)
+	}
+
+	reqs := s.shardQuery(r, analysis)
+
+	reqResps, err := DoRequests(ctx, s.next, reqs, s.limits)
+	if err != nil {
+		return nil, err
+	}
+
+	resps := make([]Response, 0, len(reqResps))
+	for _, reqResp := range reqResps {
+		resps = append(resps, reqResp.Response)
+	}
+
+	response, err := s.merger.MergeResponse(resps...)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func (s shardQueryByMiddleware) shardQuery(r Request, analysis querysharding.QueryAnalysis) []Request {
+	reqs := make([]Request, s.numShards)
+	for i := 0; i < s.numShards; i++ {
+		q, err := cquerysharding.InjectShardingInfo(r.GetQuery(), &storepb.ShardInfo{
```

Comment:
What about parsing the query first and parse the expr into this function? This should avoid parsing the query multiple times.

Reply:
If i parse only one time i will have to keep modifying the labels from the previous shard  - for the first shard i will add and for now on, i will need to update the shard label.

I think this will make the code way more complex and less readable, WDYT?

Reply:
Yeah, looks like we have to parse it everytime to modify the labels